source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -437,6 +456,65 @@ int ContextImpl::verifyCallback(X509_STORE_CTX* store_ctx, void* arg) {
 }
 
+enum ssl_verify_result_t ContextImpl::customVerifyCallback(SSL* ssl, uint8_t* out_alert) {
+  auto* extended_socket_info = reinterpret_cast<Envoy::Ssl::SslExtendedSocketInfo*>(
+      SSL_get_ex_data(ssl, ContextImpl::sslExtendedSocketInfoIndex()));
+  if (extended_socket_info->certificateValidationResult().has_value()) {
+    if (extended_socket_info->certificateValidationResult().value() ==
+        Ssl::ValidateResult::Pending) {
+      return ssl_verify_retry;
+    }
+    ENVOY_LOG(debug, "Already has a result: {}",
+              static_cast<int>(extended_socket_info->certificateValidationResult().value()));
+    // Already has a binary result, return immediately.
+    *out_alert = extended_socket_info->tlsAlert();
+    return ValidationResultToSslVerifyResult(
+        extended_socket_info->certificateValidationResult().value());
+  }
+  // Hasn't kicked off any validation for this connection yet.
+  SSL_CTX* ssl_ctx = SSL_get_SSL_CTX(ssl);
+  ContextImpl* context_impl = static_cast<ContextImpl*>(SSL_CTX_get_app_data(ssl_ctx));
+  auto* transport_socket_option =
+      static_cast<const Network::TransportSocketOptions*>(SSL_get_app_data(ssl));
+  Ssl::ValidateResult result =
+      context_impl->customVerifyCertChain(extended_socket_info, transport_socket_option, ssl,
+                                          /*error_details=*/nullptr, out_alert);
+  return ValidationResultToSslVerifyResult(result);
+}
+
+Ssl::ValidateResult
+ContextImpl::customVerifyCertChain(Envoy::Ssl::SslExtendedSocketInfo* extended_socket_info,
+                                   const Network::TransportSocketOptions* transport_socket_options,
+                                   SSL* ssl, std::string* error_details, uint8_t* out_alert) {
+  ASSERT(Runtime::runtimeFeatureEnabled("envoy.reloadable_features.tls_aync_cert_validation"));
+  STACK_OF(X509)* cert_chain = SSL_get_peer_full_cert_chain(ssl);
+  if (cert_chain == nullptr) {
+    *out_alert = SSL_AD_INTERNAL_ERROR;
+    if (extended_socket_info) {
+      extended_socket_info->setCertificateValidationStatus(
+          Envoy::Ssl::ClientValidationStatus::NotValidated);
+    }
+    stats_.fail_verify_error_.inc();
+    ENVOY_LOG(debug, "verify cert failed: no cert chain");
+    return Ssl::ValidateResult::Failed;
+  }
+  const char* name = nullptr;
+  size_t name_len = 0;
+  SSL_get0_ech_name_override(ssl, &name, &name_len);
+  absl::string_view ech_name_override(name, name_len);
+  ASSERT(cert_validator_);
+  // Do not provide async callback here, but defer its creation to extended_socket_info if the
+  // vadlidation is async.
+  Ssl::ValidateResult result = cert_validator_->doCustomVerifyCertChain(
+      *cert_chain, nullptr, extended_socket_info, transport_socket_options, SSL_get_SSL_CTX(ssl),
+      ech_name_override, SSL_is_server(ssl), error_details, out_alert);
+  if (result != Ssl::ValidateResult::Pending && extended_socket_info != nullptr) {
```

Comment:
seems `extended_socket_info` can't be nullptr. otherwise the code here doesn't check the nullptr https://github.com/envoyproxy/envoy/pull/21417/files#diff-2b39f212c2f760ae547351a46f92ad08c0ec84c556d678ebe9388b9d008e7313R462